### PR TITLE
JIT: look for escaping byrefs

### LINKED ISSUE #16210: Invalid x64 codegen with fixed statement and struct
Hey,

I found an invalid codegen (both hitting CoreCLR and .NET Framework).  For CoreCLR, I'm testing it with netcoreapp2.0 (my current dotnet version is 2.1.4)

This is the offending code (make sure to compile it with unsafe and x64):

```
using System;

namespace TestBugFixedStruct
{
    public struct float4
    {
        public float4(float x, float y, float z, float w)
        {
            this.x = x;
            this.y = y;
            this.z = z;
            this.w = w;
        }


        public float x;
        public float y;
        public float z;
        public float w;

        public unsafe float this[int index]
        {
            get
            {
                fixed (float* array = &x) { return array[index]; }
            }
            set
            {
                fixed (float* array = &x) { array[index] = value; }
            }
        }
    }


    class Program
    {
        public static float Process(int i)
        {
            var test = new float4(5, 4, 3, 2);
            return test[i];
        }

        static void Main(string[] args)
        {
            var method = typeof(Program).GetMethod("Process");
            for (int i = 0; i < 4; i++)
            {
                Console.WriteLine($"Result: {method.Invoke(null, new object[] { i })}");
            }
        }
    }
}
```
It should print: 

``` 
Result: 5
Result: 4
Result: 3
Result: 2
```
but instead it prints garbage like this (for index > 0)

```
Result: 5
Result: 0.08975968
Result: 1.905766E-43
Result: -1.469787E-33
```
Looking at the memory, it seems that the index calculation is correct, but the constructor on the stack (`new float4(5,4,3,2)`) is getting corrupted/or partial (haven't followed exactly what is happening in the details). Only the value `5` is correctly layout in memory, but not the following values.

If I change the body of the indexer to something less efficient (like `if index == 0 return x;`...etc.) it is working. I don't mean that it is the offending part. It looks more like the conjunction of the two (constructor and indexer) getting squashed in the codegen.

Let me know if you need more details.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
If a byref is passed to a call or used in an assign, assume it escapes and
that subsequent dereferences may be made at arbitrary offsets.

This handles cases where a user takes the address of one field of a struct,
then uses that byref that to access other fields or non-field content within
the struct, and the byref formation and use are split across statements or
across caller-callee contexts.

Closes #16210.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@erozenfeld @briansull PTAL
cc @dotnet/jit-contrib 

There may be other operations we need to handle this way too.

Desktop testing underway, will report back if anything turns up there.

jit-diffs results shows modest impact:
```
Total bytes of diff: -4760 (-0.02% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file regressions by size (bytes):
         125 : NuGet.Versioning.dasm (0.46% of base)
          80 : NuGet.Packaging.Core.Types.dasm (0.90% of base)
          66 : NuGet.RuntimeModel.dasm (0.44% of base)
          39 : NuGet.Packaging.Core.dasm (0.57% of base)
          36 : NuGet.DependencyResolver.Core.dasm (0.13% of base)

Top file improvements by size (bytes):
        -625 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.03% of base)
        -581 : System.Private.CoreLib.dasm (-0.02% of base)
        -536 : System.Private.Xml.dasm (-0.02% of base)
        -306 : Microsoft.CodeAnalysis.CSharp.dasm (-0.01% of base)
        -260 : Microsoft.CodeAnalysis.dasm (-0.03% of base)

102 total files with size differences (89 improved, 13 regressed), 28 unchanged.

Top method regessions by size (bytes):
          52 : NuGet.Versioning.dasm - NuGet.Versioning.VersionComparer:GetHashCode(ref):int:this
          40 : NuGet.Versioning.dasm - NuGet.Versioning.VersionRangeComparer:GetHashCode(ref):int:this
          38 : System.Private.Xml.dasm - System.Xml.XmlTextReaderImpl:ParseAttributeValueSlow(int,ushort,ref):this
          32 : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationReader:ParseArrayType(ref):struct:this
          30 : NuGet.RuntimeModel.dasm - NuGet.RuntimeModel.RuntimeDescription:GetHashCode():int:this

Top method improvements by size (bytes):
         -90 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.Scanner:ScanDateLiteral(struct):ref:this
         -76 : System.Runtime.Numerics.dasm - System.Numerics.Complex:Sqrt(struct):struct
         -72 : System.Private.CoreLib.dasm - System.Diagnostics.Tracing.EventProvider:WriteEvent(byref,long,long,long,ref):bool:this
         -68 : System.Private.Xml.dasm - System.Xml.Schema.XsdDateTime:op_Implicit(struct):struct (2 methods)
         -66 : Microsoft.VisualBasic.dasm - Microsoft.VisualBasic.CompilerServices.OverloadResolution:InsertIfMethodAvailable(ref,ref,int,bool,ref,int,ref,ref,bool,ref,ref)

2453 total methods with size differences (2403 improved, 50 regressed), 152548 unchanged.
```

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -19217,6 +19217,32 @@ Compiler::fgWalkResult Compiler::fgMarkAddrTakenLocalsPreCB(GenTree** pTree, fgW
         // them as AXC_IndWide.
         //
 
+        case GT_CALL:
+        {
+            // Scan for byref args
+            GenTreeCall* const call = tree->AsCall();
+
+            if (call->gtCallObjp)
```
Comment (mikedn) [inline src/jit/morph.cpp]: Isn't this a bit too conservative? Surely people won't call a struct method and then expect that they can access memory past `this`?

Comment (AndyAyersMS) [reply src/jit/morph.cpp]: Yeah, that probably is too conservative.

**Comment by AndyAyersMS** [comment]
Update jit-diff stats:
```
Total bytes of diff: -4407 (-0.02% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file regressions by size (bytes):
          23 : System.Reflection.Metadata.dasm (0.03% of base)

Top file improvements by size (bytes):
        -602 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.03% of base)
        -545 : System.Private.CoreLib.dasm (-0.01% of base)
        -359 : Microsoft.CodeAnalysis.CSharp.dasm (-0.02% of base)
        -254 : Microsoft.CodeAnalysis.dasm (-0.03% of base)
        -181 : System.Net.Http.dasm (-0.06% of base)

85 total files with size differences (84 improved, 1 regressed), 45 unchanged.

Top method regessions by size (bytes):
          38 : System.Private.Xml.dasm - System.Xml.XmlTextReaderImpl:ParseAttributeValueSlow(int,ushort,ref):this
          32 : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationReader:ParseArrayType(ref):struct:this
          24 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionEvaluator:PerformCompileTimeBinaryOperation(ushort,byte,ref,ref,ref):ref
          22 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - <>c__DisplayClass42_0:<FastSerialization.IFastSerializable.FromStream>b__0():this
          21 : System.Reflection.Metadata.dasm - System.Reflection.Metadata.TypeDefinition:GetCustomAttributes():struct:this

Top method improvements by size (bytes):
         -90 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.Scanner:ScanDateLiteral(struct):ref:this
         -76 : System.Runtime.Numerics.dasm - System.Numerics.Complex:Sqrt(struct):struct
         -72 : System.Private.CoreLib.dasm - System.Diagnostics.Tracing.EventProvider:WriteEvent(byref,long,long,long,ref):bool:this
         -66 : Microsoft.VisualBasic.dasm - Microsoft.VisualBasic.CompilerServices.OverloadResolution:InsertIfMethodAvailable(ref,ref,int,bool,ref,int,ref,ref,bool,ref,ref)
         -57 : System.Net.Http.dasm - <StartRequest>d__105:MoveNext():this

2142 total methods with size differences (2125 improved, 17 regressed), 152859 unchanged.
```

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -19158,11 +19158,9 @@ Compiler::fgWalkResult Compiler::fgMarkAddrTakenLocalsPreCB(GenTree** pTree, fgW
                 // We may need to Quirk the storage size for this LCL_VAR
                 // some PInvoke signatures incorrectly specify a ByRef to an INT32
                 // when they actually write a SIZE_T or INT64
-                if (axc == AXC_Addr)
-                {
-                    comp->gtCheckQuirkAddrExposedLclVar(tree, fgWalkPre->parentStack);
-                }
+                comp->gtCheckQuirkAddrExposedLclVar(tree, fgWalkPre->parentStack);
```
Comment (briansull) [inline src/jit/morph.cpp]: This is the main change that implements, Right?
"assume it escapes and that subsequent dereferences may be made at arbitrary offsets."

Comment (briansull) [reply src/jit/morph.cpp]: Plus you added support to look into at Assignements and arguments to Calls properly.

Comment (AndyAyersMS) [reply src/jit/morph.cpp]: The main change is the one in the callback for `GT_CALL` and `GT_ASG`.

That change alters the address context that reaches here so I had to broaden this condition.

**Pull request merged**